### PR TITLE
Step7: タスクを登録・更新・削除できるようにしよう

### PR DESCRIPTION
**Pull request by ShihaoXu0420**
## Requirements
- タスクの一覧画面、作成画面、詳細画面、編集画面を作成しましょう
  - `rails generate` コマンドでコントローラとビューを作成します
  - コントローラとビューに必要な実装を追加しましょう
  - 作成、更新、削除後はそれぞれflashメッセージを画面に表示させましょう
- `routes.rb` を編集して、 `http://localhost:3000/` でタスクの一覧画面が表示されるようにしましょう
- `Gemfile` で [RuboCop](https://github.com/rubocop/rubocop)（Rubyの静的コード解析ツール） か [fablicop](https://github.com/Fablic/fablicop)をインストールしましょう。設定やコマンド実行方法は各READMEから確認しましょう。

## Done
- ### index page and delete func
  - index endpoint: GET `/`
  - delete endpoint: DELETE `/:id`
image

- ### detail page
  - endpoint: GET `/:id`
image

- ### new/create page
  - new endpoint: GET `/new`
  - create endpoint: POST `/`
image

- ### edit/update page
  - edit endpoint: GET `/:id/edit`
  - update endpoint: PATCH `/:id`
image

### CODE REVIEW

File: myapp/.gitignore [hunk 1]
```diff
@@ -6,4 +6,5 @@
 yarn-debug.log*
 .yarn-integrity
 tmp/
-log/
\ No newline at end of file
+log/
+.byebug_history
```
Comment (ksak19) [inline myapp/.gitignore:10]: [nits] Recommend turn this on
image

File: myapp/app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def new
+  end
+
+  def create
+    @new_task = Task.new(params.require(:task).permit(:title, :description))
+    @new_task.save
+      
+    flash[:success] = "New task was created successfully!"
```
Comment (ksak19) [inline myapp/app/controllers/tasks_controller.rb]: Did you run `bundle exec rubocop -a`? Anyway, single quotation is preferable
```suggestion
    flash[:success] = 'New task was created successfully!'
```

File: myapp/app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    redirect_to action: "index"
+  end
+
+  def show
+    @task = Task.find(params[:id])
```
Comment (ksak19) [inline myapp/app/controllers/tasks_controller.rb]: There are the same codes in other lines. Let's think about how to make it DRY in the following steps

File: myapp/app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  def edit
+    @task = Task.find(params[:id])
+  end
+
+  def update
+    @task = Task.find(params[:id])
+    @task.update(params.require(:task).permit(:title, :description))
+
+    flash[:notice] = "Edit task successfully!"
+    redirect_to action: "index"
```
Comment (ksak19) [inline myapp/app/controllers/tasks_controller.rb]: This is more common
```suggestion
    redirect_to tasks_path
```

File: myapp/app/views/tasks/edit.html.erb
Comment (ksak19) [inline myapp/app/views/tasks/edit.html.erb:1]: indent looks like four spaces but it's better to use two spaces. VSCodeの設定でTab Sizeを変えられるので見てみてください！

File: myapp/config/routes.rb [hunk 5]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  scope(path_name: {index: ""}) do
+    resources :tasks, path: "/"
+  end
```
Comment (ksak19) [inline myapp/config/routes.rb]: Simpleにこれでいいかもです
```suggestion
  root 'tasks#index'
  resources :tasks
```
https://railsguides.jp/routing.html#root%E3%82%92%E4%BD%BF%E3%81%86

**Review by ksak19 — COMMENTED** [review]
Left some comments! Please check them out!

File: myapp/config/routes.rb (hunk 5 above)
Comment (ShihaoXu0420) [reply myapp/config/routes.rb]: As you can see the pic below, there will be a `tasks/` prefix in the url, and in this training's case I don't think this prefix is necessary so I removed it intentionally.

image

Comment (ksak19) [reply myapp/config/routes.rb]: I understand your intention! Let's see if we really don't need `/tasks` in path names 😄

File: myapp/app/views/tasks/edit.html.erb
Comment (ksak19) [reply myapp/app/views/tasks/edit.html.erb:1]: (This is just example comment. Just reference)
Tab sizeを変更しました！
https://github.com/Fablic/training/pull/1628/commits/390d1fee9f970ad169957ea0e46b2decc3db8415

**Review by ksak19 — APPROVED** [review]
Thank you for the quick fix! Looks good to me for now!

From next time, 

- please re-request review so the reviewer get notifications after you fix something. Usually we don't notify on slack about PR review since everyone turn on the notification when they get any requests 😄 
  image
- When you get any comments on your codes and finish fixing it, please reply to each comment thread with commit hash (I put a comment as example). It's easier for reviewer to re-review!
  https://github.com/Fablic/training/pull/1628#discussion_r1602388157

I know this is just simple Rails training but I want you to experience the actual process of team development in our team

File: myapp/app/controllers/tasks_controller.rb (hunk 4 above)
Comment (ShihaoXu0420) [reply myapp/app/controllers/tasks_controller.rb]: (test)
Refactored controller
https://github.com/Fablic/training/pull/1628/commits/423d43431d4b86f72897f8f350490a35adcbf751

**Pull request merged**